packages-exp/auth-compat-exp/src/auth.ts

```diff
@@ -0,0 +1,269 @@
+    return (this as unknown) as externs.Auth;
+  }
+
+  applyActionCode(code: string): Promise<void> {
+    return impl.applyActionCode(this._asExtern(), code);
+  }
+
+  checkActionCode(code: string): Promise<compat.ActionCodeInfo> {
+    return impl.checkActionCode(this._asExtern(), code);
+  }
+
+  confirmPasswordReset(code: string, newPassword: string): Promise<void> {
+    return impl.confirmPasswordReset(this._asExtern(), code, newPassword);
+  }
+
+  async createUserWithEmailAndPassword(
+    email: string,
+    password: string
+  ): Promise<compat.UserCredential> {
+    return convertCredential(
+      impl.createUserWithEmailAndPassword(this._asExtern(), email, password)
+    );
+  }
+  // currentUser: User | null;
+  fetchSignInMethodsForEmail(email: string): Promise<string[]> {
+    return impl.fetchSignInMethodsForEmail(this._asExtern(), email);
+  }
+  isSignInWithEmailLink(emailLink: string): boolean {
+    return impl.isSignInWithEmailLink(this._asExtern(), emailLink);
+  }
+  async getRedirectResult(): Promise<compat.UserCredential> {
+    const credential = await impl.getRedirectResult(
+      this._asExtern(),
+      impl.browserPopupRedirectResolver
+    );
+    if (!credential) {
+      return {
+        credential: null,
+        user: null
+      };
+    }
+    return convertCredential(Promise.resolve(credential!));
+  }
+  // languageCode: string | null;
+  //       settings: AuthSettings;
+  onAuthStateChanged(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    nextOrObserver: Observer<any> | ((a: compat.User | null) => any),
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    error?: (error: compat.Error) => any,
+    completed?: Unsubscribe
+  ): Unsubscribe {
+    return super._onAuthStateChanged(
+      nextOrObserver as externs.NextOrObserver<externs.User | null>,
+      error as ErrorFn,
+      completed
+    );
+  }
+  onIdTokenChanged(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    nextOrObserver: Observer<any> | ((a: compat.User | null) => any),
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    error?: (error: compat.Error) => any,
+    completed?: Unsubscribe
+  ): Unsubscribe {
+    return super._onIdTokenChanged(
+      nextOrObserver as externs.NextOrObserver<externs.User | null>,
+      error as ErrorFn,
+      completed
+    );
+  }
+  sendSignInLinkToEmail(
+    email: string,
+    actionCodeSettings: compat.ActionCodeSettings
+  ): Promise<void> {
+    return impl.sendSignInLinkToEmail(
+      this._asExtern(),
+      email,
+      actionCodeSettings
+    );
+  }
+  sendPasswordResetEmail(
+    email: string,
+    actionCodeSettings?: compat.ActionCodeSettings | null
+  ): Promise<void> {
+    return impl.sendPasswordResetEmail(
+      this._asExtern(),
+      email,
+      actionCodeSettings || undefined
+    );
+  }
+  async setPersistence(persistence: string): Promise<void> {
+    function convertPersistence(
```

Comment:
why does this need to be a separate function?

Reply:
switch statements don't return their value in TS, this allows me to be functional and avoid mutable variables, happy to change if you have strong opinions

Reply:
ah that's fine